From c045cbc913c2da2e3924084c4559fb43bb9f711b Mon Sep 17 00:00:00 2001
From: Kevin McCarthy <kevin@8t8.us>
Date: Sun, 1 Mar 2020 18:05:46 -0800
Subject: [PATCH] Add %i to $background_format.

This will show the parent message id for replies and forwarded
messages.
---
 background.c | 13 ++++++++++++-
 init.h       |  1 +
 2 files changed, 13 insertions(+), 1 deletion(-)

diff --git a/background.c b/background.c
index 3d8af3c0..3b115f14 100644
--- a/background.c
+++ b/background.c
@@ -380,13 +380,24 @@ static const char *bg_format_str (char *dest, size_t destlen, size_t col,
                                   unsigned long data, format_flag flags)
 {
   BG_ENTRY *entry = (BG_ENTRY *)data;
-  HEADER *hdr = entry->process->sctx->msg;
+  SEND_CONTEXT *sctx = entry->process->sctx;
+  HEADER *hdr = sctx->msg;
   char tmp[SHORT_STRING];
   char buf[LONG_STRING];
+  const char *msgid;
   int optional = (flags & MUTT_FORMAT_OPTIONAL);
 
   switch (op)
   {
+    case 'i':
+      msgid = sctx->cur_message_id;
+      if (!msgid && sctx->tagged_message_ids)
+        msgid = sctx->tagged_message_ids->data;
+      if (!optional)
+        mutt_format_s (dest, destlen, fmt, NONULL (msgid));
+      else if (!msgid)
+        optional = 0;
+      break;
     case 'n':
       snprintf (tmp, sizeof (tmp), "%%%sd", fmt);
       snprintf (dest, destlen, tmp, entry->num);
diff --git a/init.h b/init.h
index 26ee16ca..9e9df5eb 100644
--- a/init.h
+++ b/init.h
@@ -405,6 +405,7 @@ struct option_t MuttVars[] = {
   ** menu.  The following \fCprintf(3)\fP-style sequences are
   ** understood:
   ** .dl
+  ** .dt %i .dd parent message id (for replies and forwarded messages)
   ** .dt %n .dd the running number on the menu
   ** .dt %p .dd pid of the $$editor process
   ** .dt %r .dd comma separated list of ``To:'' recipients
-- 
2.26.2

